src/mlpack/core/data/csv_parser_impl.hpp

```diff
@@ -0,0 +1,202 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------
+*/
+#ifndef MLPACK_CORE_DATA_CSV_PARSER_IMPL_HPP
+#define MLPACK_CORE_DATA_CSV_PARSER_IMPL_HPP
+
+#include "csv_parser.hpp"
+
+namespace mlpack
+{
+namespace data
+{
+
+  /**
+   * Given the address of a martix element(val)
+   * sets it equal to the provided value(token)
+   * example calling: convert_token<eT>(x.at(row, col), token)
+   */
+  template<typename MatType>
+  bool ConvertToken(typename MatType::elem_type& val, const std::string& token)
+  {
+    const size_t N = size_t(token.length());
+
+    if (N == 0)
+    {
+      val = typename MatType::elem_type(0);
+
+      return true;
+    }
+
+    const char* str = token.c_str();
+
+    // checking for nan, +inf, -inf
+    // in both upper and lower case
+    // using arma::Datum which basically
+    // contains all the physical constants
+    if ((N == 3) || (N == 4))
+    {
+      const bool neg = (str[0] == '-');
+      const bool pos = (str[0] == '+');
+
+      const size_t offset = ((neg || pos) && (N == 4)) ? 1 : 0;
+
+      const char sig_a = str[offset];
+      const char sig_b = str[offset+1];
+      const char sig_c = str[offset+2];
+
+      if (((sig_a == 'i') || (sig_a == 'I')) &&
+          ((sig_b == 'n') || (sig_b == 'N')) &&
+	  ((sig_c == 'f') || (sig_c == 'F')))
+      {
+        // val = if(neg == true) ? -INF : +INF
+        val = neg ? -(std::numeric_limits<typename MatType::elem_type>::infinity()) :
+	            std::numeric_limits<typename MatType::elem_type>::infinity();
+
+        return true;
+      }
+      else if (((sig_a == 'n') || (sig_a == 'N')) &&
+	       ((sig_b == 'a') || (sig_b == 'A')) &&
+	       ((sig_c == 'n') || (sig_c == 'N')))
+      {
+        val = std::numeric_limits<typename MatType::elem_type>::quiet_NaN();
+
+        return true;
+      }
+    }
+
+    char* endptr = nullptr;
+
+    if ((std::is_floating_point<typename MatType::elem_type>::value) ||
+        (std::is_integral<typename MatType::elem_type>::value))
+    {
+      val = typename MatType::elem_type(std::strtod(str, &endptr));
+    }
+    else
+    {
+      if (std::is_signed<typename MatType::elem_type>::value)
+      {
+        val = typename MatType::elem_type(std::strtoll(str, &endptr, 10));
+      }
+      else
+      {
+        if (str[0] == '-')
+        {
+          val = typename MatType::elem_type(0);
+          return true;
+        }
+
+        val = typename MatType::elem_type( std::strtoull(str, &endptr, 10));
+      }
+    }
+
+    if (str == endptr)
+    {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Loads the data from the csv file
+   * into the give MatType
+   */
+  template<typename MatType>
+  bool LoadCSVV(MatType& x, std::fstream& f, std::string&)
```

Comment:
```suggestion
  bool LoadCSVV(MatType& x, std::fstream& f, std::string&)
```
What does the `std::string&` do here? and why it is added? would you elaborate on this one thanks